components/sup/src/manager/service/pipe_hook_client.rs

```diff
@@ -0,0 +1,319 @@
+                              .ok();
+
+        // We want to wait until we know the named pipe is up and running before returning OK
+        // If we suspect anything is wrong with the pipe we should terminate the pwsh process
+        let start = Instant::now();
+        loop {
+            match self.pipe_exists() {
+                Ok(_) => return Ok(()),
+                Err(err) => {
+                    if start.elapsed() >= Duration::from_secs(START_PIPE_TIMEOUT_SECONDS) {
+                        self.win32_result(unsafe {
+                                processthreadsapi::TerminateProcess(handle.raw(), 1)
+                            })?;
+                        return Err(Error::NamedPipeTimeoutOnStart(service_group.to_string(),
+                                                                  self.hook_name.to_string(),
+                                                                  err));
+                    }
+                }
+            }
+            thread::sleep(Duration::from_millis(10));
+        }
+    }
+
+    // Simple call to win32 API to see if named pipe exists.
+    // Note that we cannot use Path::exists for named pipes
+    fn pipe_exists(&self) -> io::Result<()> {
+        let pipe_path: Vec<u16> = OsStr::new(&self.abs_pipe_name()).encode_wide()
+                                                                   .chain(once(0))
+                                                                   .collect();
+        self.win32_result(unsafe { namedpipeapi::WaitNamedPipeW(pipe_path.as_ptr(), 10) })?;
+        Ok(())
+    }
+
+    fn quit(&self) -> Result<()> {
+        if self.pipe_exists().is_ok() {
+            debug!("Telling {} pipe server to quit", self.pipe_name);
+            let (mut pipe, poll) = self.connect()?;
+            self.pipe_ready(&poll, Ready::writable())?;
+            pipe.write_all(&SIGNAL_QUIT)?;
+        }
+        Ok(())
+    }
+
+    fn win32_result(&self, i: i32) -> io::Result<i32> {
+        if i == 0 {
+            Err(io::Error::last_os_error())
+        } else {
+            Ok(i)
+        }
+    }
+}
+
+impl Drop for PipeHookClient {
+    fn drop(&mut self) {
+        if let Err(err) = self.quit() {
+            error!("Unable to tell {} pipe server to quit. {}",
+                   self.pipe_name, err);
+        }
+    }
+}
+
+fn stream_output<T>(out: Option<T>, log_file: &PathBuf, preamble_str: &str)
+    where T: Read
+{
+    let mut log = File::create(&log_file).unwrap_or_else(|_| {
+                                             panic!("{}: couldn't create log output file",
+                                                    preamble_str)
+                                         });
+
+    if let Some(o) = out {
+        for line in BufReader::new(o).lines() {
+            if let Ok(ref l) = line {
+                outputln!(preamble preamble_str, l);
+                log.write_fmt(format_args!("{}\n", l))
```

Comment:
OPTIONAL: this can be simplified a bit:
```suggestion
                writeln!(log, "{}", l)
```